prost-derive/src/lib.rs

```diff
@@ -206,6 +221,18 @@ fn try_message(input: TokenStream) -> Result<TokenStream, Error> {
             }
         }
 
+        impl #impl_generics ::prost::MessageDescriptor for #ident #ty_generics #where_clause {
+            fn message_name(&self) -> &'static str {
+                stringify!(#ident)
+            }
+            fn package_name(&self) -> &'static str {
+                #pkg_name
+            }
+            fn type_url(&self) -> &'static str {
+                #type_url
+            }
+        }
```

Comment:
I believe the above methods should be associated functions instead. Someone will be able to use `Foo::type_url()` instead of constructing `Foo` and then calling the method.

For example this is how pack/unpack to `Any` will be implemented with associated functions:

```rust
fn pack_to_any<M>(msg: M) -> prost_types::Any
where
    M: prost::Message + prost::MessageDescriptor,
{
    prost_types::Any {
        type_url: M::type_url().to_owned(),
        value: msg.encode_to_vec(),
    }
}

fn unpack_from_any<M>(msg: prost_types::Any) -> Option<M>
where
    M: prost::Message + prost::MessageDescriptor + Default,
{
    if msg.type_url == M::type_url() {
        Some(M::decode(&msg.value[..]).ok()?)
    } else {
        None
    }
}
```